### PR TITLE
Clean Up Some Data Structures a Bit

### PR DESCRIPTION
**Pull request by CodaFi**
Shuffle the docs around on some decls, drop some unused decls, and prettify the interface to BidirectionalMap with some new docs.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test macOS

**Review by artemcm — APPROVED** [review]

**Comment by artemcm** [comment]
@swift-ci test

File: Sources/SwiftDriver/IncrementalCompilation/BidirectionalMap.swift [hunk 1]
```diff
@@ -63,40 +64,53 @@ public struct BidirectionalMap<T1: Hashable, T2: Hashable>: Equatable, Sequence
...
+  /// the subscript operator to perform an in-place update.
+  ///
+  /// - Parameters:
+  ///   - v: The new value to add to the two-level map.
+  ///   - key: The two-level key to associate with value.
+  /// - Returns: The value that was replaced, or `nil` if a new key-value pair was added.
+  @discardableResult
   public mutating func updateValue(_ newValue: T2, forKey key: T1) -> T2? {
-    let oldValue = map1.updateValue(newValue, forKey: key)
-    _ = oldValue.map {map2.removeValue(forKey: $0)}
```
Comment (davidungar) [inline Sources/SwiftDriver/IncrementalCompilation/BidirectionalMap.swift:76]: I think the loss of this line introduces a bug. Luckily, the bug won't be triggered in the status quo.

**Review by davidungar — COMMENTED** [review]
Looks good!

**Pull request merged**